src/pybind/mgr/dashboard/frontend/src/app/ceph/new-dashboard/dashboard/dashboard.component.ts

```diff
@@ -124,4 +130,20 @@ export class DashboardComponent implements OnInit, OnDestroy {
       });
     });
   }
+
+  preparePgStatus() {
+    this.categoryPgAmount = {};
+    let pgs = 0;
+    this.healthService.getMinimalHealth().subscribe((data: any) => {
+      _.forEach(data.pg_info.statuses, (pgAmount, pgStatesText) => {
+        const categoryType = this.pgCategoryService.getTypeByStates(pgStatesText);
+        if (_.isUndefined(this.categoryPgAmount[categoryType])) {
+          this.categoryPgAmount[categoryType] = 0;
+        }
+        this.categoryPgAmount[categoryType] += pgAmount;
+        pgs += pgAmount;
+        this.totalPgs = pgs;
+      });
+    });
+  }
```

Comment:
I'd avoid doing kind of calculation in the front-end. And if not, I'd go for a pipe for this, since Angular automatically caches pipes outputs for an unchanged input:
```html
<div *ngIf="healthService.minimalHealth$ | async as minimalHealth"
{{ minimalHealth.pg_info.statuses | cdPgInfoToWhatever }}
```

That said, Angular Components should have minimal logic/data processing. Just presentation-related code. Everything else should be buried into services. IMHO Component TS file should just define the Component decorator/metadata, the dependencies and that's all. Everything else should be done in the Component HTML Template via async pipe (and other pipes). And Services should provide ready to use (by the template) data, either front-end Services or, even better, back-end ones.

Reply:
I used a pipe to process this.